yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,1048 @@
+   *         if it has finished.
+   */
+  def isPostThreadActive: Boolean = {
+    postThreadActive.get
+  }
+
+  /**
+   * The YARN application ID of this history service
+   * @return the application ID provided when the service started
+   */
+  def applicationId: ApplicationId =  _applicationId
+
+  /**
+   * The YARN attempt ID of this history service
+   * @return the attempt ID provided when the service started
+   */
+  def attemptId: Option[ApplicationAttemptId] =  _attemptId
+
+  /**
+   * Reset the timeline client
+   * <p>
+   * 1. Stop the timeline client service if running.
+   * 2. set the `timelineClient` field to `None`
+   */
+  def stopTimelineClient(): Unit = {
+    _timelineClient.foreach(_.stop())
+    _timelineClient = None
+  }
+
+  /**
+   * Create the timeline domain.
+   *
+   * A Timeline Domain is a uniquely identified 'namespace' for accessing parts of the timeline.
+   * Security levels are are managed at the domain level, so one is created if the
+   * spark acls are enabled. Full access is then granted to the current user,
+   * all users in the configuration options `"spark.modify.acls"` and `"spark.admin.acls"`;
+   * read access to those users and those listed in `"spark.ui.view.acls"`
+   *
+   * @return an optional domain string. If `None`, then no domain was created.
+   */
+  private def createTimelineDomain(): Option[String] = {
+    val sparkConf = sparkContext.getConf
+    val aclsOn = sparkConf.getBoolean("spark.ui.acls.enable",
+        sparkConf.getBoolean("spark.acls.enable", false))
+    if (!aclsOn) {
+      logDebug("ACLs are disabled; not creating the timeline domain")
+      return None
+    }
+    val predefDomain = sparkConf.getOption(TIMELINE_DOMAIN)
+    if (predefDomain.isDefined) {
+      logDebug(s"Using predefined domain $predefDomain")
+      return predefDomain
+    }
+    val current = UserGroupInformation.getCurrentUser.getShortUserName
+    val adminAcls  = stringToSet(sparkConf.get("spark.admin.acls", ""))
+    val viewAcls = stringToSet(sparkConf.get("spark.ui.view.acls", ""))
+    val modifyAcls = stringToSet(sparkConf.get("spark.modify.acls", ""))
+
+    val readers = (Seq(current) ++ adminAcls ++ modifyAcls ++ viewAcls).mkString(" ")
+    val writers = (Seq(current) ++ adminAcls ++ modifyAcls).mkString(" ")
+    var domain = DOMAIN_ID_PREFIX + _applicationId
+    logInfo(s"Creating domain $domain with readers: $readers and writers: $writers")
+
+    // create the timeline domain with the reader and writer permissions
+    val timelineDomain = new TimelineDomain()
+    timelineDomain.setId(domain)
+    timelineDomain.setReaders(readers)
+    timelineDomain.setWriters(writers)
+    try {
+      timelineClient.putDomain(timelineDomain)
+      Some(domain)
+    } catch {
+      case e: Exception => {
+        logError(s"cannot create the domain $domain", e)
+        // fallback to default
+        None
```

Comment:
security in ATS is best described as "weak". For the timeline support, the history server has to have access to all the histories saved to ATS, and, as it doesn't have a SPNEGO front end, it's not keeping its information secret. My own stance (this code predates me working on the patch) is that trying to lock down the ATS data here isn't worth trying. 

If it is, then maybe the entity name used to save & retrieve data should be something customizable -different user groups could then publish securely, with separate history server instances needed to access the different dataset